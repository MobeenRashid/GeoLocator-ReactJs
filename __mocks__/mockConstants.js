import { arrayToMap } from "../src/utils/arrayUtils";

const __mockLocationList = [
    {
        id: '1',
        address: 'Berlin',
        lng: "123",
        lat: "456"
    },
    {
        id: '2',
        address: 'Munich',
        lng: "124",
        lat: "457"
    }
];


const __mockState = {
    locations: __mockLocationList,
    requestingLocations: false
}

export { __mockState, __mockLocationList };