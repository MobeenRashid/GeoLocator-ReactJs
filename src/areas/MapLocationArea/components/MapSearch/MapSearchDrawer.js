import React from 'react';
import { Drawer } from 'antd';
import MapLocationSearch from './MapLocationSearch';
import BasicMap from '../../../../components/GoogleMap/BasicMap';
import { GOOGLE_MAP_URL } from '../../../../common/constants';
import DrawerFooter from './DrawerFooter';

const MapSearchDrawer = (props) => {
    return (
        <Drawer
            className="search-drawer"
            title="Search Location"
            placement="top"
            onClose={props.onClose}
            visible={props.visible}
            height={window.innerHeight}
        >
            <div className="drawer-content">
                <MapLocationSearch disabled={props.disabled} reset={!props.visible} onLocationSelect={props.onLocationSelect} defaultLocation={props.selectedLocation}></MapLocationSearch>
                <BasicMap
                    isMarkerShown
                    locations={props.locations}
                    googleMapURL={GOOGLE_MAP_URL}
                    loadingElement={<div className="map-loading-div" style={{ height: '100%' }} />}
                    containerElement={<div className="map-container" style={{ height: '100%' }} />}
                    mapElement={<div className="map-element" style={{ height: '100%' }} />}
                >
                </BasicMap>
            </div>
            <DrawerFooter {...props}></DrawerFooter>
        </Drawer>
    );
}

export default MapSearchDrawer;
