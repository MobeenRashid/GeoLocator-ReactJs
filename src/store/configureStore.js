import { createStore, applyMiddleware } from "redux";
import thunk from 'redux-thunk';

import appReducer from "../reducers/appReducer";

const configureStore = (initialState) => createStore(appReducer, initialState, applyMiddleware(thunk));
export default configureStore;