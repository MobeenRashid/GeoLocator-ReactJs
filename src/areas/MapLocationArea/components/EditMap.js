import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import LocationApi from '../../../api/LocationApi';
import { updateLocation } from '../../../actions/locationActions';
import { Button, message } from 'antd';
import MapSearchDrawer from './MapSearch/MapSearchDrawer';


export class EditMap extends Component {
    constructor(props) {
        super(props);
        this.state = {
            selectedLocation: null,
            redirectToReturn: false
        }

        this.locationApi = new LocationApi();

        this.onUpdateMap = this.onUpdateMap.bind(this);
        this.onClose = this.onClose.bind(this);
        this.onLocationSelect = this.onLocationSelect.bind(this);
    }

    componentDidMount() {
        let { match, locationData } = this.props;
        let selectedLocation = locationData.find(loct => loct.id === match.params.id);
        if (!selectedLocation) {
            message.warn('Oops! looks like select location not exist');
            this.setState({ redirectToReturn: true });
        }
        this.setState({ selectedLocation });
    }

    onLocationSelect(location) {
        let { selectedLocation } = this.state;
        this.setState({
            selectedLocation: {
                id: selectedLocation && selectedLocation.id,
                lat: location.lat,
                lng: location.lng,
                address: location.address
            }
        });
    }

    onClose() {
        this.setState({ redirectToReturn: true });
    }

    onUpdateMap() {
        let { selectedLocation } = this.state
        let { dispatch, locationData } = this.props;

        if (!selectedLocation) {
            message.warn('Please select a location to add');
            return;
        }

        if (!locationData.find(loct => loct.id === selectedLocation.id)) {
            message.warn('Oops! looks like selected location not exist');
            return;
        }

        dispatch(updateLocation(selectedLocation));
        this.setState({
            selectedLocation: null,
            redirectToReturn: true,
        });
    }

    handleRedirect() {
        let { redirectToReturn } = this.state;
        if (redirectToReturn) {
            return <Redirect to={this.props.returnUrl || '/locations'}></Redirect>
        }
    }

    render() {
        let { selectedLocation } = this.state;
        return (
            <div className="edit-location">
                <MapSearchDrawer
                    visible={true}
                    locations={[selectedLocation]}
                    selectedLocation={selectedLocation}
                    onLocationSelect={this.onLocationSelect}
                    onClose={this.onClose}
                    footerAction={<Button onClick={this.onUpdateMap} type="primary">Update Map</Button>}
                />
                {this.handleRedirect()}
            </div>
        );
    }
}

const mapStateToProps = state => {
    return {
        locationData: state.locations
    }
}

export default connect(mapStateToProps)(EditMap);