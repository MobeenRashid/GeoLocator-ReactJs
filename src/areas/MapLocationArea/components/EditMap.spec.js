import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import { render, mount } from 'enzyme';
import { Provider } from 'react-redux';
import axios from 'axios';
import configureStore from '../../../store/configureStore'
import { __mockState, __mockLocationList } from '../../../../__mocks__/__mockState';
import MapSearchDrawer from './MapSearch/MapSearchDrawer';
import EditMap, { mapStateToProps } from './EditMap';

jest.mock('axios');

describe('EditMap Component', () => {

    test('Should map state to props correctly', () => {
        var componentProps = mapStateToProps(__mockState);
        expect(componentProps.locationData).toEqual(__mockState.locations);
    });

    test("Should render correctly", () => {
        const store = configureStore(__mockState);
        const component = render(<Provider store={store}>
            <BrowserRouter>
                <EditMap />
            </BrowserRouter>
        </Provider>);
        expect(component).toMatchSnapshot();
    });

    test("Should update location correctly", () => {
        const store = configureStore(__mockState);
        const component = mount(<Provider store={store}>
            <BrowserRouter>
                <EditMap />
            </BrowserRouter>
        </Provider>);

        let __potsDam = __mockLocationList[0];
        __potsDam.address = 'Potsdam';

        component.find(MapSearchDrawer).props().onLocationSelect(__potsDam)

        axios.put.mockImplementation(() => Promise.resolve({ data: __potsDam }));
        component.find('#btnUpdateLocation').first().simulate('click');

        const lastState = store.getState();
        const location = lastState.locations.find(loct => loct && loct.id === __potsDam.id);
        expect(location).toBe(__potsDam);

        component.unmount();
    });
});