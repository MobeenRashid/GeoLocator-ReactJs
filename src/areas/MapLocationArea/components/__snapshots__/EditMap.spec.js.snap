// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditMap Component Should render correctly 1`] = `
<div
  class="edit-location"
/>
`;
